### PR TITLE
Worker submit performance improvements

### PR DESCRIPTION
**Pull request by cosa65**
# Description

# Details
#### URL to issue
https://github.com/hms-dbmi-cellenics/issues/issues/17

#### Link to staging deployment URL (or set N/A)
N/A

#### Integration test branch
master

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.

### Code updates
Have best practices and ongoing refactors being observed in this PR
- [ ] Migrated any selector / reducer used to the new format.

### Manual/unit testing
- [ ] Tested changes using InfraMock locally **or** no tests required for change, e.g. Kubernetes chart updates.
- [ ] Validated that current unit tests for code work as expected and are sufficient for code coverage **or** no unit tests required for change, e.g. documentation update.
- [ ] Unit tests written **or** no unit tests required for change, e.g. documentation update.

### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.

- [ ] Started end-to-end tests on the latest commit.

### Documentation updates
- [ ] Relevant Github READMEs updated **or** no GitHub README updates required.
- [ ] Relevant Wiki pages created/updated **or** no Wiki updates required.

### Optional
- [ ] Staging environment is unstaged before merging.
- [ ] Photo of a cute animal attached to this PR.

### CODE REVIEW

File: python/src/worker/response.py [hunk 1]
```diff
@@ -30,25 +32,47 @@ def _construct_data_for_upload(self):
                 zipfile.write(self.result.data)
             else:
                 info('Encoding and compressing json work result')
-                json.dump(self.result.data, zipfile)
+                ujson.dump(self.result.data, zipfile)
 
 
+
+        gz_body_bytes = None
+        # If size is less than 2 mb, then send it over notification too
```
Comment (cosa65) [inline python/src/worker/response.py]: "Send it directly over socket"*

File: python/src/worker/response.py [hunk 2]
```diff
@@ -30,25 +32,47 @@ def _construct_data_for_upload(self):
...
+        # Needs to be done here because upload_fileobj closes the file:
+        # https://github.com/boto/boto3/issues/929
+        mb = 1000000
+        body_size = sys.getsizeof(gzipped_body)
+        info(f"Body size is {body_size}")
+        if (body_size <= 2 * mb):
```
Comment (cosa65) [inline python/src/worker/response.py]: I still need to test an fine-tune this cutoff a bit to make sure we switch to sending the `signedUrl` before the data becomes too big for the socket (sockets are not fast for transmitting large pieces of data)

Comment (kafkasl) [reply python/src/worker/response.py]: I am not very much in favor of these kind of fine-tunings. It looks like they depend too much on network speed so how do you plan to find an "optimal" value?

File: python/src/worker/response.py (hunk 1 above)
Comment (kafkasl) [reply python/src/worker/response.py]: I think we should have a design doc discussing this. It greatly changes the current flow of data and I'd like to see if it's worth the increased complexity of the resulting code.

File: python/src/worker/response.py (hunk 2 above)
Comment (kafkasl) [reply python/src/worker/response.py]: interesting and relevant answer in stack overflow https://stackoverflow.com/a/12978658.

I think the things to keep in mind are:
1. would all the small request be sent sequentially? 
2. is sending through socket blocking? 
3. is sending to socket faster / slower (from the executing code point of view) than upload to S3?
4. would the message be sent to multiple UIs if you have the experiment opened in many tabs?

Comment (cosa65) [reply python/src/worker/response.py]: Thanks for the link! It definitely brings up some interesting points. I would say to keep in mind though that we are not using just sockets, but rather a wrapper for redis's pub/sub mechanism, you can check that we use a library called `socket_io_emitter`. My understanding is that it does end up as a socketio message when reaching the ui though.

I definitely think 2mb is way too much, I had just set a big value to begin the comparison on. The actual value we'll settle on will probably be way smaller.

> I am not very much in favor of these kind of fine-tunings. It looks like they depend too much on network speed so how do you plan to find an "optimal" value?

I am saying fine-tuning, but not going overboard with this at all. I am performing some tests both from my place (SA) and am planning to get help from someone else in europe too to compare whether the gain is still worth it from a place closer to the server itself. I'll update here with the stats I pick up.

1. I didn't understand this question
2. It isn't blocking. I believe there is a blocking emit which cna be used, but we are not.
3. From the informal tests I had done originally, which prompted me to implement this, it does make a difference on small enough messages (in average, from south america, it makes 1 sec of difference). The question of approximately how small the message should be is what I was planning to use some stats for.
4. No, only if all tabs are performing the same work request at the same time. The only messages that arrive to all open tabs of an experiment in particualr are the ones sent to the `ExperimentUpdates-{experimentId}` socket.

Comment (cosa65) [reply python/src/worker/response.py]: Related reply [here](https://github.com/hms-dbmi-cellenics/ui/pull/885#discussion_r1327549119)

**Review by ivababukova — APPROVED** [review]

**Pull request merged**